tools/tune/eq/gui.m

```diff
@@ -0,0 +1,476 @@
+  axis([log2(nyquist_f / (2 ^ n_octaves)), log2(nyquist_f), -24, 18]);
+  for c = 1:2
+    for i = 1:length(positions)
+      EQ = initialize_biquad(i, c, positions(i, 1:end), main_panes, Fs, EQ);
+    endfor
+  endfor
+  % octave is silly and doesn't support references so we have to re-write the
+  % control matrix so the callbacks have access without using globals
+  for c = 1:2
+    for i = 1:length(positions)
+      update_control_elements(EQ(i, c), EQ);
+    endfor
+  endfor
+  fn = figure(fn, "toolbar", "none", "menubar", "none", 'Position', [200 200 300 190]); fn = fn + 1;
+  menu_window = struct();
+  menu_window.ip_box = uicontrol( ...
+        "style", "edit", ...
+        "position", [120 160 180 30]);
+  menu_window.ip_label = uicontrol( ...
+        "style", "text",
+        "string", "DUT IP:", ...
+        "position", [0 160 120 30]);
+  menu_window.device_box = uicontrol( ...
+        "style", "edit", ...
+        "position", [120 130 180 30]);
+  menu_window.device_label = uicontrol( ...
+        "style", "text",
+        "string", "Device Num:", ...
+        "position", [0 130 120 30]);
+  menu_window.alsa_box = uicontrol( ...
+        "style", "edit", ...
+        "position", [120 100 180 30]);
+  menu_window.alsa_label = uicontrol( ...
+        "style", "text",
+        "string", "ALSA Ctl Num:", ...
+        "position", [0 100 120 30]);
+  menu_window.upload = uicontrol( ...
+        "style", "pushbutton", ...
+        'string', 'Push config to DUT', ...
+        "callback", @push_config, ...
+        "position", [0 50 300 50], ...
+        "userdata", struct("menu_window", menu_window, "EQ", EQ));
+  menu_window.load_dsp = uicontrol( ...
+        "style", "pushbutton", ...
+        'string', 'Load dsp.ini', ...
+        "callback", @eq_load_dsp_ini, ...
+        "position", [0 0 300 50], ...
+        "userdata", struct("EQ", EQ));
+end
+
+function conf = get_config()
+  % since octave has no definitive way to constants we are going to hack a
+  % struct into behaving like one
+  conf = struct( ...
+      "Fs", 48000 ...
+      );
+end
+
+function set_udata_field(h, field, value)
+  udata = get(h, "userdata");
+  udata = setfield(udata, field, value);
+  set(h, "userdata", udata);
+end
+
+function peq = build_peq(channel, EQ)
+  peq = [];
+  for i = 1:8
+    biquad = EQ(i, channel);
+    if get(biquad.enable_switch, "value") == 1
+      type = get(biquad.type.dropdown, "value");
+      freq = str2num(get(biquad.freq.edit, "string"));
+      Q = str2double(get(biquad.Q.edit, "string"));
+      gain = str2double(get(biquad.gain.edit, "string"));
+      % dropdown maps straight to filter index
+      % filter only uses needed arguments, all others will be ignored
+      peq = [peq; type, freq, gain, Q];
+    endif
+  endfor
+end
+
+function update_plot(channel, main_panes, EQ)
+  Fs = get_config().Fs;
+  nyquist_f = Fs / 2;
+  pl = main_panes(channel).plot;
+  spectrum_size = 4096;
+  n = linspace(0, nyquist_f, spectrum_size);
```

Comment:
This could be logspace() with less data points for quicker UI since the UI is using logarithmic scale. E.g. logspace(log10(20), log10(20e3), 1000) creates a logarithmic 1000 points grid.

Reply:
Definitely good point, will update this and try it out. I remember initially I was having problems with octave overflowing its gui components, I will try this out and hopefully optimize this.